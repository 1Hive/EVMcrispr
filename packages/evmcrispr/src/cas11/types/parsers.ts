import type { Parser } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

import type { Node } from './ast';

export type NodeParser<T = Node> = Parser<T, string, NodeParserState>;
