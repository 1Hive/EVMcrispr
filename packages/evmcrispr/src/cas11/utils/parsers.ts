import type { Err } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

export const buildParserError = (
  { index, error }: Err<string, null> | Err<string, NodeParserState>,
  type: string,
  msg?: string,
): string => {
  const splitRes = error.split('got');
  const wrongValueEncountered =
    splitRes.length === 2 ? splitRes[1].trim() : null;

  const errorMessage = msg
    ? `${msg}${wrongValueEncountered ? `, got ${wrongValueEncountered}` : ''}`
    : error.split('): ')[1];
  return `${type}(col: ${index}): ${errorMessage}`;
};

export const getIncorrectReceivedValue = (errorMsg: string): string => {
  const splitRes = errorMsg.split('got ');

  if (splitRes.length === 2) {
    return `, got ${splitRes[1].trim()}`;
  }

  return '';
};
