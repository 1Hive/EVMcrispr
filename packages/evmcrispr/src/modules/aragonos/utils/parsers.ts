import type { PublicClient } from "viem";
import { parseAbiItem } from "viem";

import type { Address } from "../../../types";
import type { ParsedApp, Repo } from "../types";
import { getSystemApp, isSystemApp } from "./interfaces";

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return "";
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : "";
};

const fetchImplementationAddress = (
  appAddress: Address,
  client: PublicClient,
): Promise<Address> => {
  return client.readContract({
    address: appAddress,
    abi: [
      parseAbiItem("function implementation() public view returns (address)"),
    ],
    functionName: "implementation",
  });
};

export const parseApp = async (
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  app: any,
  client: PublicClient,
): Promise<ParsedApp> => {
  const { address, appId, roles, version } = app;
  const { name: repoName, registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress:
      codeAddress ?? (await fetchImplementationAddress(address, client)),
    contentUri,
    name,
    registryName: registry?.name,
    roles: (roles as any[]).map((role) => ({
      ...role,
      roleHash: role["hash"] ?? role["roleHash"],
    })),
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};
