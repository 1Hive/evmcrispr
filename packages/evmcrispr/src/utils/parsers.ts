import { getSystemApp, isSystemApp } from '.';
import type { ParsedApp, Repo } from '../types';

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return '';
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : '';
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseApp = (app: any): ParsedApp => {
  const { address, appId, implementation, repoName, roles, version } = app;
  const { registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress: codeAddress ?? implementation.address,
    contentUri,
    name,
    registryName: registry?.name,
    roles,
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};

export const timeUnits: { [key: string]: number } = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  mo: 2592000,
  y: 31536000,
};
