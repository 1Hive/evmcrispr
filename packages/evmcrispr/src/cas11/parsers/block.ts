import type { Parser } from 'arcsecond';
import { char, coroutine, recursiveParser } from 'arcsecond';

import type { BlockExpressionNode } from '../types';
import { NodeType } from '../types';
import { commandExpressionParser } from './command';
import {
  endOfLine,
  optionalEmptyLines,
  optionalWhitespace,
  surroundedBy,
} from './utils';

export const blockExpressionParser: Parser<BlockExpressionNode, string, any> =
  recursiveParser(() =>
    coroutine(function* () {
      yield char('(');

      yield optionalWhitespace;

      yield endOfLine;

      const scopedCommands = yield optionalEmptyLines(commandExpressionParser);

      yield surroundedBy(optionalWhitespace)(char(')'));

      return {
        type: NodeType.BlockExpression,
        body: scopedCommands,
      } as unknown as BlockExpressionNode;
    }),
  );
