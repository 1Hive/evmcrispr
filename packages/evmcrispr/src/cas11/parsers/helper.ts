import { char, coroutine, letters, possibly, recursiveParser } from 'arcsecond';

import type { StringLiteralNode } from '../types';
import { NodeType } from '../types';
import { argumentsParser } from './expression';

export const helperFunctionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('@');

    const name = yield letters.map(
      (value): StringLiteralNode => ({
        type: NodeType.StringLiteral,
        value,
      }),
    );

    const args = yield possibly(argumentsParser);

    return {
      type: NodeType.HelperFunctionExpression,
      name,
      args: args === null ? [] : args,
    };
  }),
);
