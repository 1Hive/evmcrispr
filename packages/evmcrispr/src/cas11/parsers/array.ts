import type { Parser } from 'arcsecond';
import {
  between,
  char,
  coroutine,
  recursiveParser,
  sequenceOf,
} from 'arcsecond';

import type { ArrayExpressionNode } from '../types';
import { NodeType } from '../types';

import { argumentExpressionParser } from './expression';

import { commaSeparated, optionalWhitespace } from './utils';

export const arrayExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    between(sequenceOf([char('['), optionalWhitespace]))(
      sequenceOf([optionalWhitespace, char(']')]),
    )(commaSeparated(argumentExpressionParser)).map(
      (elements: any): ArrayExpressionNode => {
        return {
          type: NodeType.ArrayExpression,
          elements,
        };
      },
    ),
);

export const arrayExpresionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('[');
    yield optionalWhitespace;
    const elements = yield commaSeparated(argumentExpressionParser);
    yield optionalWhitespace;
    yield char(']');

    return {
      type: NodeType.ArrayExpression,
      elements,
    };
  }),
);
