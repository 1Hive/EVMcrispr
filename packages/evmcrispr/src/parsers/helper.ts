import {
  char,
  coroutine,
  lookAhead,
  possibly,
  recursiveParser,
  regex,
  takeLeft,
} from 'arcsecond';

import type {
  ArgumentExpressionNode,
  HelperFunctionNode,
  NodeParser,
} from '../types';
import { NodeType } from '../types';
import { buildParserError } from '../utils/parsers';
import { argumentsParser } from './expression';
import {
  callOperatorParser,
  comma,
  createNodeLocation,
  enclosingLookaheadParser,
  locate,
  openingCharParser,
} from './utils';

export const HELPER_PARSER_ERROR = 'HelperParserError';
const helperNameParser = takeLeft(regex(/^(?!-|\.)[a-zA-Z\-.]+(?<!-|\.)/))(
  enclosingLookaheadParser([
    char('('),
    comma,
    char(']'),
    callOperatorParser,
    char(')'),
  ]),
).errorMap((err) =>
  buildParserError(err, HELPER_PARSER_ERROR, 'Expecting a helper name'),
);

export const helperFunctionParser: NodeParser<HelperFunctionNode> =
  recursiveParser(() =>
    locate<HelperFunctionNode>(
      coroutine(run => {
        run(char('@'));

        const name: string | undefined = run(helperNameParser);

        let args: ArgumentExpressionNode[] = [];

        if (run(possibly(lookAhead(openingCharParser('('))))) {
          args = run(argumentsParser.errorMap((err) =>
            buildParserError(err, HELPER_PARSER_ERROR),
          ));
        }

        return [name, args];
      }),
      ({ data, index, result: [initialContext, [name, args]] }) => ({
        type: NodeType.HelperFunctionExpression,
        name: name as HelperFunctionNode['name'],
        args: args as HelperFunctionNode['args'],
        loc: createNodeLocation(initialContext, {
          line: data.line,
          index,
          offset: data.offset,
        }),
      }),
    ),
  );
