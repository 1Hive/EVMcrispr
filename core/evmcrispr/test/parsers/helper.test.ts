import type { Case } from '@1hive/evmcrispr-test-common';
import { runParser, runParserError } from '@1hive/evmcrispr-test-common';

import { helperFunctionParser } from '../../src/parsers/helper';

describe.concurrent('Parsers - helper functions', () => {
  describe.each<Case>([
    {
      title: '',
      value: '@token(WETH)',
    },
    {
      title: 'with call expression',
      value:
        '@helperFunction(anotherToken::symbol(), "this is a string param", 10e18)',
    },
    {
      title: 'with no arguments',
      value: '@now',
    },
    {
      title: 'with nested helpers',
      value: `@token('DAI', @calc(34, @innerHelper(true)))`,
    },
  ])('', ({ title, value }) => {
    it(`should parse helper ${title} correctly`, () => {
      const parsedValue = runParser(helperFunctionParser, value);

      expect(parsedValue).toMatchSnapshot();
    });
  });

  describe.each<Case>([
    { title: 'an invalid name', value: '@asd&$6' },
    { title: 'no closing parenthesis', value: '@helper(asda,1e18' },
    { title: 'no opening parenthesis', value: '@helperarg1, 1e18, ,)' },
    { title: 'empty arguments', value: '@helper(arg1, 1e18, ,)' },
  ])('', ({ title, value }) => {
    it(`should fail when parsing a helper with ${title}`, () => {
      const error = runParserError(helperFunctionParser, value);

      expect(error).toMatchSnapshot();
    });
  });
});
