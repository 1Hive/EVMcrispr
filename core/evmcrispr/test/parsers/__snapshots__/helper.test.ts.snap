// Vitest Snapshot v1

exports[`Parsers - helper functions > should fail when parsing a helper with an invalid name 1`] = `"HelperParserError(1:4): Expecting a helper name, got '&$6...'"`;

exports[`Parsers - helper functions > should fail when parsing a helper with empty arguments 1`] = `"HelperParserError(1:20): Expecting a valid expression, got ','"`;

exports[`Parsers - helper functions > should fail when parsing a helper with no closing parenthesis 1`] = `"HelperParserError(1:17): Expecting character ')', but got end of input."`;

exports[`Parsers - helper functions > should fail when parsing a helper with no opening parenthesis 1`] = `"HelperParserError(1:10): Expecting a helper name, got '1, 1e...'"`;

exports[`Parsers - helper functions > should parse helper  correctly 1`] = `
{
  "args": [
    {
      "loc": {
        "end": {
          "col": 11,
          "line": 1,
        },
        "start": {
          "col": 7,
          "line": 1,
        },
      },
      "type": "ProbableIdentifier",
      "value": "WETH",
    },
  ],
  "loc": {
    "end": {
      "col": 12,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "name": "token",
  "type": "HelperFunctionExpression",
}
`;

exports[`Parsers - helper functions > should parse helper with call expression correctly 1`] = `
{
  "args": [
    {
      "args": [],
      "loc": {
        "end": {
          "col": 38,
          "line": 1,
        },
        "start": {
          "col": 16,
          "line": 1,
        },
      },
      "method": "symbol",
      "target": {
        "loc": {
          "end": {
            "col": 28,
            "line": 1,
          },
          "start": {
            "col": 16,
            "line": 1,
          },
        },
        "type": "ProbableIdentifier",
        "value": "anotherToken",
      },
      "type": "CallExpression",
    },
    {
      "loc": {
        "end": {
          "col": 64,
          "line": 1,
        },
        "start": {
          "col": 40,
          "line": 1,
        },
      },
      "type": "StringLiteral",
      "value": "this is a string param",
    },
    {
      "loc": {
        "end": {
          "col": 71,
          "line": 1,
        },
        "start": {
          "col": 66,
          "line": 1,
        },
      },
      "power": 18,
      "type": "NumberLiteral",
      "value": "10",
    },
  ],
  "loc": {
    "end": {
      "col": 72,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "name": "helperFunction",
  "type": "HelperFunctionExpression",
}
`;

exports[`Parsers - helper functions > should parse helper with nested helpers correctly 1`] = `
{
  "args": [
    {
      "loc": {
        "end": {
          "col": 12,
          "line": 1,
        },
        "start": {
          "col": 7,
          "line": 1,
        },
      },
      "type": "StringLiteral",
      "value": "DAI",
    },
    {
      "args": [
        {
          "loc": {
            "end": {
              "col": 22,
              "line": 1,
            },
            "start": {
              "col": 20,
              "line": 1,
            },
          },
          "type": "NumberLiteral",
          "value": "34",
        },
        {
          "args": [
            {
              "loc": {
                "end": {
                  "col": 41,
                  "line": 1,
                },
                "start": {
                  "col": 37,
                  "line": 1,
                },
              },
              "type": "BoolLiteral",
              "value": true,
            },
          ],
          "loc": {
            "end": {
              "col": 42,
              "line": 1,
            },
            "start": {
              "col": 24,
              "line": 1,
            },
          },
          "name": "innerHelper",
          "type": "HelperFunctionExpression",
        },
      ],
      "loc": {
        "end": {
          "col": 43,
          "line": 1,
        },
        "start": {
          "col": 14,
          "line": 1,
        },
      },
      "name": "calc",
      "type": "HelperFunctionExpression",
    },
  ],
  "loc": {
    "end": {
      "col": 44,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "name": "token",
  "type": "HelperFunctionExpression",
}
`;

exports[`Parsers - helper functions > should parse helper with no arguments correctly 1`] = `
{
  "args": [],
  "loc": {
    "end": {
      "col": 4,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "name": "now",
  "type": "HelperFunctionExpression",
}
`;
